doc: document Sopel configuration

Exirel opened the pull request:
This PR replaces #1564 because at some point I just... wrote the full documentation for the configuration file (464 new lines). So here it is. You can check in #1564 that I've modified the documentation according to all the recent changes (or I hope so).

This is a big change, and it's mostly English. So it needs editing, and a lot of patience and tolerance, because to be honest, I'm terrible at this.

@dgw have fun with that. :wink:  (and sorry)

dgw commented on sopel/cli/run.py:
```diff
@@ -113,17 +113,21 @@ def signal_handler(sig, frame):
+                        help=(
+                            "Daemonize Sopel. "
+                            "(deprecated, and will be removed in Sopel 8; "
+                            "use ``sopel start -d`` instead)"))
```
I thought the deprecation note was left off of this intentionally. https://github.com/sopel-irc/sopel/pull/1493/files#r262470506

Exirel replied:
I don't remember exactly what I did. What I remember is that, from the moment `sopel start` exists and is available in master, `sopel -d` should be deprecated.

dgw replied:
Back in that PR, [I asked](https://github.com/sopel-irc/sopel/pull/1493/files#r262285696) about `-d`, "I'm guessing this has no deprecation note because e.g. sopel -d will work identically in Sopel 8, when start becomes the default action?" You replied, "Exactly!"

Unless that's no longer the case, this deprecation warning should be removed. (If it is no longer the case, you should open another PR to add the deprecation warning at runtime like the others.)

Exirel replied:
Ha! Silly me.

dgw commented on sopel/cli/run.py:
```diff
@@ -196,18 +205,30 @@ def build_parser():
         default=False,
-        dest='modules')
+        dest='modules',
+        help='Check for Sopel plugins that requires configuration, and run '
```
```suggestion
        help='Check for Sopel plugins that require configuration, and run '
```

dgw commented on sopel/cli/run.py:
```diff
@@ -196,18 +205,30 @@ def build_parser():
-        dest='modules')
+        dest='modules',
+        help='Check for Sopel plugins that requires configuration, and run '
+             'their configuration wizard.')
```
```suggestion
             'their configuration wizards.')
```

dgw commented on sopel/cli/run.py:
```diff
@@ -196,18 +205,30 @@ def build_parser():
+        description='Stop a running Sopel instance. '
+                    'This command determines the instance to quit by the name '
+                    'of the configuration file used ("default", or the one '
+                    'from the ``-c/--config`` option). '
```
```suggestion
                    'from the ``-c``/``--config`` option). '
```

dgw commented on sopel/cli/utils.py:
```diff
@@ -222,7 +222,11 @@ def add_common_arguments(parser):
-        help='Use a specific configuration file')
+        help='Use a specific configuration file. '
+             'A config name can be given and the configuration file will be '
+             'find in Sopel\'s homedir (default to ``~/.sopel/default.cfg``). '
```
```suggestion
             'found in Sopel\'s homedir (defaults to ``~/.sopel/default.cfg``). '
```

Also, doesn't the option actually default to `default`, rather than that full path?

Exirel replied:
The option itself defaults to `None` indeed, but its usage is to default to "the default" config path.

dgw commented on sopel/config/__init__.py:
```diff
@@ -1,15 +1,43 @@
-object is initialized.
+"""Sopel's configuration module.
+
+The :class:`sopel.config.Config` object provides an interface to access Sopel's
```
```suggestion
The :class:`~sopel.config.Config` object provides an interface to access Sopel's
```

dgw commented on sopel/config/__init__.py:
```diff
@@ -1,15 +1,43 @@
+
+The :class:`sopel.config.Config` object provides an interface to access Sopel's
+configuration file. It exposes the configuration's sections through its
+attributes as object, and these objects expose their directives through their
```
```suggestion
attributes as objects, which in turn expose their directives through *their*
```

dgw commented on sopel/config/__init__.py:
```diff
@@ -1,15 +1,43 @@
+        bacon = config.types.ValidatedAttribute('bacon')
+
+The ``[core]`` section itself is represented by the
+:class:`sopel.config.core_section.CoreSection`, which is a subclass of
```
```suggestion
:class:`~sopel.config.core_section.CoreSection` class, which is a subclass of
```

dgw commented on sopel/config/__init__.py:
```diff
@@ -1,15 +1,43 @@
+The ``[core]`` section itself is represented by the
+:class:`sopel.config.core_section.CoreSection`, which is a subclass of
+:class:`~sopel.config.types.StaticSection`. It is automatically added when
+the :class:`Config` object is instanciated; it uses
```
```suggestion
the :class:`Config` object is instantiated; it uses
```

dgw commented on sopel/config/core_section.py:
```diff
@@ -166,7 +184,7 @@ class CoreSection(StaticSection):
 
     help_prefix = ValidatedAttribute('help_prefix', default='.')
-    """The prefix to use in help"""
+    """The prefix to use in help."""
```
```suggestion
    """The prefix to use in help output."""
```

Reads cleaner.

dgw commented on sopel/config/core_section.py:
```diff
@@ -235,34 +264,42 @@ def homedir(self):
 
     nick_auth_password = ValidatedAttribute('nick_auth_password')
-    """The password to use to authenticate your nick."""
+    """The password to use to authenticate your nick.
```
```suggestion
    """The password to use to authenticate the bot's nick.
```

"your" is confusing, isn't it?

Exirel commented on docs/source/cli.rst:
```diff
@@ -0,0 +1,37 @@
+
+The ``configure`` sub-command will run a brief configuration wizard to set up
+the basic options: the bot's nick, its admin's nick, the server and channels
+to join. By default, it creates the file ``~/.sopel/default.cfg``; creating
```
```suggestion
to join. By default, it creates the file ``~/.sopel/default.cfg`` (creating
```

Exirel commented on docs/source/cli.rst:
```diff
@@ -0,0 +1,37 @@
+The ``configure`` sub-command will run a brief configuration wizard to set up
+the basic options: the bot's nick, its admin's nick, the server and channels
+to join. By default, it creates the file ``~/.sopel/default.cfg``; creating
+the homedir directory if necessary.
```
```suggestion
the homedir directory if necessary.)
```

Exirel commented on docs/source/configure.rst:
```diff
@@ -0,0 +1,464 @@
+    channels = #sopel
+
+which tells the bot what is its name and its owner, and which server to
+connect to and which channels to join.
```
Sometimes I forgot to add it, but I swear I'm an Oxford partisan!

Exirel commented on docs/source/configure.rst:
```diff
@@ -0,0 +1,464 @@
+User Modes
+----------
+
+To configure User modes, the :attr:`~CoreSection.modes` can be used:
```
I thought as well about linking to some IRC wiki page, but couldn't settle on one in particular. I'm sure we'll figure this out eventually.

Exirel commented on docs/source/configure.rst:
```diff
@@ -0,0 +1,464 @@
+.. code-block:: ini
+
+   [core]
+   mode = BpR
```
lol I suck

Exirel commented on docs/source/configure.rst:
```diff
@@ -0,0 +1,464 @@
+Ping timeout
+------------
+
+By default, if Sopel doesn't get a Ping from the server every 120s, it will
```
I love it.

Exirel commented on docs/source/configure.rst:
```diff
@@ -0,0 +1,464 @@
+
+* default port to connect to IRC will be 6697
+* certificate will be verified if :attr:`~CoreSection.verify_ssl` is set to
+  true, and its location can be configured with :attr:`~CoreSection.ca_certs`.
```
Yeah, I wasn't sure about that part, at all. I don't like dealing with ca certs and stuff, and I usually left that to any ops I can reach.

Exirel commented on docs/source/configure.rst:
```diff
@@ -0,0 +1,464 @@
+   Sopel uses the built-in :func:`ssl.wrap_socket` function to wrap the socket
+   used for the IRC connection.
+
+   Certificate is required if and only if :attr:`~CoreSection.verify_ssl` is
```
ca certs? see my previous comment: I just don't like that part. Someone else should write the doc about it. :stuck_out_tongue_closed_eyes:

Exirel commented on docs/source/configure.rst:
```diff
@@ -0,0 +1,464 @@
+Channels
+--------
+
+By default, Sopel won't connect to any channels. The list of channels to
```
??? why did I write that?

Exirel commented on docs/source/configure.rst:
```diff
@@ -0,0 +1,464 @@
+--------
+
+By default, Sopel won't connect to any channels. The list of channels to
+connect to is configured by :attr:`~CoreSection.channels`:
```
```suggestion
join is configured by :attr:`~CoreSection.channels`:
```

Exirel commented on docs/source/configure.rst:
```diff
@@ -0,0 +1,464 @@
+
+And for **nick-based** methods:
+
+* ``nickserv``,
```
```suggestion
* ``nickserv``
```

Exirel commented on docs/source/configure.rst:
```diff
@@ -0,0 +1,464 @@
+And for **nick-based** methods:
+
+* ``nickserv``,
+* ``authserv``,
```
```suggestion
* ``authserv``
```

Exirel commented on docs/source/configure.rst:
```diff
@@ -0,0 +1,464 @@
+
+* ``nickserv``,
+* ``authserv``,
+* ``Q``,
```
```suggestion
* ``Q``
```

Exirel commented on docs/source/configure.rst:
```diff
@@ -0,0 +1,464 @@
+Nick-based
+..........
+
+When :attr:`~CoreSection.nick_auth_method` is defined, the configuration
```
```suggestion
When :attr:`~CoreSection.nick_auth_method` is defined, the settings
```

Exirel commented on docs/source/configure.rst:
```diff
@@ -0,0 +1,464 @@
+.. code-block:: ini
+
+   [core]
+   exclude = reload, meetbot
```
Easier to read by a human, and readability matters. I'm in favor of keeping these spaces.

dgw replied:
Fair enough, as long as it won't break anything if a user actually includes a space as shown.

Exirel commented on docs/source/configure.rst:
```diff
@@ -0,0 +1,464 @@
+Ignore User
+-----------
+
+To ignore users based on their hosts and/or nicks, you can use these options:
```
Oh, I ran out of ideas. I think this can wait for another PR. :grin:

Exirel commented on docs/source/configure.rst:
```diff
@@ -0,0 +1,464 @@
+
+Misc
+====
+
```
At some point in another PR, if you don't mind. :grin:

Exirel commented on docs/source/configure.rst:
```diff
@@ -0,0 +1,464 @@
+section of the configuration file. See each of its attributes for its full
+description.
+
+This file can be generated from a :doc:`console wizard<cli>` using
```
Not sure I can do that right now. :-/

Exirel commented on docs/source/configure.rst:
```diff
@@ -0,0 +1,464 @@
+.. py:module:: sopel.config.core_section
+
```
I finish my review, and I do that.

Exirel commented:
@dgw I did my best with the review, and I think it's in a reasonable state right now. Not a complete documentation, yet way better than what we had so far. I can't wait to see it live when Sopel 7 will come out!

Let me know if you think everything's good enough, and I'll stash commits and rename the `configure.rst` file to something more sensible.

dgw commented on docs/source/configure.rst:
```diff
@@ -0,0 +1,504 @@
+* :attr:`~CoreSection.host`: the server's hostname. Can be a domain name
+  (like ``chat.freenode.net``) or an IP address.
+* :attr:`~CoreSection.port`: optional, the port to connect to. Usually 6697,
+  this is the default value the bot will use to connet to the server.
```
Can probably delete "this is" if you meant to emphasize what I think you did (that 6697 will be used by default).

Bonus spelling fix included for free :D

```suggestion
  the default value the bot will use to connect to the server.
```

dgw commented on docs/source/configure.rst:
```diff
@@ -0,0 +1,504 @@
+
+.. note::
+
+   Sopel will try to look at these file for the CA certs pem file required by
```
```suggestion
   Sopel will try to look at one of these files for the CA certs pem file required by
```

dgw commented on docs/source/configure.rst:
```diff
@@ -0,0 +1,504 @@
+   * ``/etc/ssl/ca-bundle.pem`` (OpenSUSE)
+
+   This is required if :attr:`~CoreSection.verify_ssl` is set to true. It is
+   possible to set this file with :attr:`~CoreSection.ca_certs`.
```
This is just one idea. :)

```suggestion
   possible to set the file used with :attr:`~CoreSection.ca_certs`. This is
   useful if e.g. Sopel cannot find the CA certs file, or you need Sopel to
   trust a CA not trusted by the system.
```

dgw commented on docs/source/configure.rst:
```diff
@@ -0,0 +1,504 @@
+   [core]
+   channels = #sopel, #sopelunkers
+
+It is possible to slow down the initial join of channels using
```
```suggestion
It is possible to slow down the initial joining of channels using
```

dgw commented on docs/source/configure.rst:
```diff
@@ -0,0 +1,504 @@
+Authentication
+==============
+
+Sopel provide two ways to authenticate: a simple method, and multi-stage
```
```suggestion
Sopel provides two ways to authenticate: a simple method, and multi-stage
```

dgw commented on docs/source/configure.rst:
```diff
@@ -0,0 +1,504 @@
+Multi-stage
+-------------
+
+In some case, an IRC bot needs to use both server-based and
```
```suggestion
In some cases, an IRC bot needs to use both server-based and
```

dgw commented on docs/source/configure.rst:
```diff
@@ -0,0 +1,504 @@
+.. versionadded:: 7.0
+
+   Using SQLAlchemy for the database has been added in Sopel 7.0, which
+   supports multiple type of databases. The configuration options required for
```
```suggestion
   supports multiple types of databases. The configuration options required for
```

dgw commented on docs/source/configure.rst:
```diff
@@ -0,0 +1,504 @@
+Alternatively, you can define a list of allowed plugins with
+:attr:`~CoreSection.enable`: plugins not in this list will be ignored. In this
+example, only the ``bugzilla`` and ``remind`` plugins are enabled (because
+the ``meetbot`` is still excluded):
```
Choose your adventure:

```suggestion
the ``meetbot`` plulgin is still excluded):
```

or

```suggestion
``meetbot`` is still excluded):
```

Exirel replied:
The latter is more friendly.

dgw commented on docs/source/configure.rst:
```diff
@@ -0,0 +1,504 @@
+.. versionchanged:: 7.0
+
+   The log filename has been renamed from ``stdio.log`` to ``<base>.stdio.log``
+   to prevent conflict when running more than one instance of Sopel.
```
```suggestion
   to prevent conflicts when running more than one instance of Sopel.
```

dgw commented on docs/source/configure.rst:
```diff
@@ -0,0 +1,504 @@
+.. versionchanged:: 7.0
+
+   The log filename has been renamed from ``raw.log`` to ``<base>.raw.log``
+   to prevent conflict when running more than one instance of Sopel.
```
```suggestion
   to prevent conflicts when running more than one instance of Sopel.
```

dgw commented on docs/source/configure.rst:
```diff
@@ -0,0 +1,504 @@
+   to prevent conflict when running more than one instance of Sopel.
+
+
+Misc
```
Let's call this section either "Misc." with the dot (it's an abbreviation), or "Miscellaneous". Or "Other". That works too (and it's short).

Exirel replied:
```suggestion
Other
```

dgw commented on sopel/config/__init__.py:
```diff
@@ -67,8 +95,14 @@ def __init__(self, filename, validate=True):
+        """The config object's associated file."""
+        basename, _ = os.path.splitext(os.path.basename(filename))
+        self.basename = basename
+        """The config's base filename, ie. the filename without the extension.
```
```suggestion
        """The config's base filename, i.e. the filename without the extension.
```

Exirel commented:
> Of special note is the conversation I "unresolved" about deprecating `-d`/`--fork` (I added a new comment there explaining why).

It's all my fault. I added a comment to explain why `-d/--fork` is not deprecated and why it shouldn't be: next time, I shouldn't do the mistake.

> The only other thing bigger than just editing is the final verdict on renaming `docs/source/configure.rst`, which I think you were going to change to `docs/source/configuration.rst`?

Exactly, if you are still OK with that.

Exirel commented on docs/source/configure.rst:
```diff
@@ -0,0 +1,505 @@
+
+
+Misc
+====
```
```suggestion
=====
```

Exirel commented:
@dgw 21 commits and 120~ comments later, I think we are very close to an end.  (I hope so)

As you said on IRC, I rebased & squash that commits. I kept the "fix" commits (one for code, one for doc), because I don't see how I could dispatch these fixes properly: the chance of a dumb mistake is too big. I know I'll mess something if I try.

What a journey! I really hope that our users will be able to understand how to configure Sopel after that. Or, at least, that we'll be able to point to the right section of the documentation when there is a question on IRC (we had a lot of them surrounding auth the past few weeks).

I think we could update (in a future PR) the `CONTRIBUTING.md` file to say something like: "if you add config options, update the doc".

Exirel commented:
I almost forgot these:

> Let's replace all instances of "sub-command" with "subcommand".

Done!

> Keep an eye out for inconsistencies in list items. Some end with a period, and some don't, sometimes within the same list. I caught some, but probably not all.

Oh yes, indeed. I spotted one last in the auth section. Fixed it!

dgw commented on docs/source/configure.rst:
```diff
@@ -0,0 +1,507 @@
+    nick = Sopel
+    host = chat.freenode.net
+    use_ssl = false
+    port = 6667
```
If you aren't planning to force-push this PR to do the `configure.rst` -> `configuration.rst` rename, ignore this. 😸

If you _will_ force-push one last time, could you update this example to give the SSL version of connecting to freenode? When users copy-and-paste this example (and they _will_), I want them to automatically get the secure version unless they specifically edit it to connect over plaintext. Sensible?

dgw commented on docs/source/configure.rst:
```diff
@@ -0,0 +1,507 @@
+by :attr:`~CoreSection.admin_accounts` or by :attr:`~CoreSection.admins`. If
+``admin_accounts`` is set, ``admins`` will be ignored.
+
+Both ``owner_account`` and ``admin_accounts`` are safer to use than a
```
Here, too, only apply this if you're already going to force-push again anyway. It's really minor.

```suggestion
Both ``owner_account`` and ``admin_accounts`` are safer to use than
```

dgw commented on docs/source/configure.rst:
```diff
@@ -0,0 +1,507 @@
+SQLite
+------
+
+There is only one options for SQLite, :attr:`~CoreSection.db_filename`, which
```
Welp. I missed this last night. 😹😿

```suggestion
There is only one option for SQLite, :attr:`~CoreSection.db_filename`, which
```

dgw commented on docs/source/configure.rst:
```diff
@@ -0,0 +1,507 @@
+Plugins
+-------
+
+By default, Sopel will load all available plugins. To exclude a plugins, you
```
…I also missed this. 😭

```suggestion
By default, Sopel will load all available plugins. To exclude a plugin, you
```

dgw commented on sopel/config/__init__.py:
```diff
@@ -1,15 +1,43 @@
+    nick = Sopel
+    host = irc.freenode.org
+    use_ssl = false
+    port = 6667
```
Same here re: secure connections to freenode by default.

dgw commented on sopel/cli/run.py:
```diff
@@ -112,18 +112,24 @@ def signal_handler(sig, frame):
     parser.add_argument("-d", '--fork', action="store_true",
-                        dest="daemonize", help="Daemonize Sopel")
+                        dest="daemonize",
+                        help=("Daemonize Sopel."))
```
It's weird that this has nested parentheses even though it's only one line. (It's also weird that this section of the code uses nested parentheses for `help` argument values when they're not needed, but changing that would require touching code that's not related to documenting the configuration.)

Maybe revert this hunk and just add the `.` in `help`?

Exirel replied:
```suggestion
                        help="Daemonize Sopel.")
```

Exirel commented:
> So close!

Will work on that by the end of the day (in my timezone). Thanks!

Exirel commented:
@dgw Sorry for the delay. :wink: 

Now, I'll become the eye of Sauron and watch over every new PR that tries to add a config option without the proper documentation.

:smiling_imp: :smiling_imp: :smiling_imp: :smiling_imp: :smiling_imp:
